File: mycollab-web/src/main/java/com/esofthead/mycollab/vaadin/mvp/PageActionChain.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.esofthead.mycollab.vaadin.mvp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 *
 * @author haiphucnguyen
 */
public class PageActionChain {
	private static Logger log = LoggerFactory.getLogger(PageActionChain.class);
	
    private List<ScreenData> chains = new ArrayList<ScreenData>();
    
    public PageActionChain(ScreenData... pageActionArr) {
        chains.addAll(Arrays.asList(pageActionArr));
        log.debug("Chain size: " + this + "---" + chains.size());
    }
    
    public PageActionChain add(ScreenData pageAction) {
        chains.add(pageAction);
        log.debug("Chain size: " + this + "---" + chains.size());
        return this;
    }
    
    public ScreenData pop() {
    	log.debug("Pop pageActionChain " + this + "---" + chains.size());
    	if (chains.size() > 0) {
    		ScreenData pageAction = chains.get(0);
            chains.remove(0);
            return pageAction;
    	} else {
    		return null;
    	}
    	
    }
    
    public ScreenData peek() {
    	ScreenData pageAction = chains.get(0);
        return pageAction;
    }
    
    public boolean hasNext() {
        return chains.size() > 0;
    }
}

File: mycollab-ui/src/main/scala/com/esofthead/mycollab/vaadin/mvp/PageActionChain.scala
package com.esofthead.mycollab.vaadin.mvp

import scala.collection.mutable

/**
  * @author MyCollab Ltd
  * @since 5.0.9
  */
class PageActionChain {
  val chains: mutable.Buffer[ScreenData[_]] = mutable.Buffer()

  def this(pageActionArr: Array[ScreenData[_]]) = {
    this()
    chains.appendAll(pageActionArr)
  }

  def this(param: ScreenData[_]) = this(Array[ScreenData[_]](param))

  def this(param1: ScreenData[_], param2: ScreenData[_]) = this(Array[ScreenData[_]](param1, param2))

  def add(pageAction: ScreenData[_]): PageActionChain = {
    chains += pageAction
    return this
  }

  def pop: ScreenData[_] = {
    if (chains.size > 0) {
      val pageAction = chains(0)
      chains.remove(0)
      return pageAction
    }
    else return null
  }

  def peek: ScreenData[_] = chains(0)

  def hasNext: Boolean = chains.length > 0
}
